### ISSUE TITLE
"feature request" show information's of prints from OctoPrint

### ISSUE THREAD

**Issue by Strankler** [prusa3d/Prusa-Firmware-Buddy#189]
**Description:**
While prints initiated from OctoPrint, the display shown no progress of the print, but only the home screen (plus temps, nozzle height, speed ratio and material).

**Feature request:**
Show same information's in the display for prints initiated from OctoPrint as for print jobs from USB drive, at least progress in percent, printing and remaining time, like the MK3 does.
Also the option to stop or pause the print would be great.

**Comment by michalxfanta** [comment]
Thank you for an interesting idea, we will consider its implementation.

**Comment by fawick** [comment]
I second this feature request.

In particular, it would be great to see support for the following codes when printing via USB cable from Octoprint
- M73
- M117

**Comment by steve220** [comment]
I'll third this request :-)

**Comment by 6ffm70** [comment]
I fourth this request

**Comment by yonie1** [comment]
+1 this would be really helpful and add value and the code to display is already there because it can display it when printing from USB.

**Comment by naltatis** [comment]
A year ago I've built this octoprint based flexi-rex progress indicator with an e-paper display on a raspberry pi. I'd love to see the magic octopus slowly getting filled on the Prusa mini screen.
<img src="https://user-images.githubusercontent.com/152287/96903621-5a71f600-1496-11eb-9afa-8da36bf8e953.jpg" width="300">

**Comment by mucek4** [comment]
Any news on this? I really miss that information and a picture of octo is useless.

**Comment by gimpfenlord** [comment]
i also "upvote" this.
currently my LCD is useless 99% of the time. so much unused potential

**Comment by Joshi425** [comment]
this would be awsome please @michalxfanta

**Comment by w90** [comment]
Great idea! Upvote!
Now in order to see the status of my print, I need to visit the octoprint.local page in browser

**Comment by LimeAyy** [comment]
Here to also request this.

**Comment by aguschanchu** [comment]
Please implement this!

**Comment by DerGolo** [comment]
+1

**Comment by Gordontgopher** [comment]
What does this give you that setting "Action Command Prompt Settings" to 'never' in Octoprint doesn't give you?

**Comment by mosal1** [comment]
+1

**Comment by thefreewheeler** [comment]
Any news on this? Hugely disappointing that the Mini's LCD shows no useful info.

**Comment by Reetro** [comment]
Any update?

**Comment by matthewlloyd** [comment]
Anyone following this issue may be interested to know that I just released custom firmware that adds support for `M73` (print progress) and `M117` (LCD messages):

http://github.com/matthewlloyd/Llama-Mini-Firmware#readme

**Comment by coaxial** [comment]
This is sorely missing. The Mini LCD is so large, why take all the space up with an Octoprint logo? It’s cumbersome to go and find a computer or phone and visit the Octoprint page just to find out how far along the print is…

**Comment by redPanther** [comment]
It's a pitty that we have a nice screen with so little information. In the meanwhile I have a workaround:
if octoprint is used, install DisplayLayerProgress Plugin. I use this for printer message: `[progress]% L=[current_layer]/[total_layers]`

then when printing, select "tune" scroll all the way down to "messages". Then you see the progress in % and current and max layer. Temperatures and z height is already displayed at the bottom. Looks absolutely not so nice as it could, but it provides necessary information during print.

**Comment by coaxial** [comment]
> Anyone following this issue may be interested to know that I just released custom firmware that adds support for `M73` (print progress) and `M117` (LCD messages):
> 
> [http://github.com/matthewlloyd/Llama-Mini-Firmware#readme](https://github.com/matthewlloyd/Llama-Mini-Firmware#readme)

Would love to try your firmware, but really not keen on voiding my warranty just to get M117 and M73 support. And there is no other way at the moment, I think.

**Comment by theHalo** [comment]
Was hoping this would have been included in today’s release. Bump.

**Comment by pat1121** [comment]
+1 would also like to see this!

**Comment by coaxial** [comment]
I have a PR (#1497) that would enable this based on the work done for the llama firmware but its not building for some reason (can't see why.) I wish someone more comfortable with C could give feedback and help pinpoint what's going on, because if the build for that PR is green then it has more chances of getting merged.

**Comment by mtaumike** [comment]
Bumping this, however it gets done.

Be fun to get some progress for when walking up to the printer, and even better, if any thumbnail that was created in the gcode file could also be shown vs 'the octopus'

**Comment by atxkev** [comment]
+1 to add this in. My MK3s screen with octoprint is more useful.

**Comment by HolgerMA** [comment]
+1

**Comment by Alshootfa** [comment]
+1 will allow us to take advantage of the Prusa Mini's awesome color display instead of covering it up.

**Comment by Aelirenn666** [comment]
I have recently started using OctoPrint with my Mini and I admit, this is one of the functions, that I miss most. I would really like to be able to see remaining print time on the display.

**Comment by Rose-David** [comment]
In my opinion, I'd rather that the Octopus logo only be used for when nothing's printing, but I can't see anything stopping the normal USB print UI from working here. Either way, I just wanna give my nice big LCD a nice big purpose.

**Comment by neo666lis** [comment]
+1
This quite annoying not having the progress status.

**Comment by coaxial** [comment]
As a reminder, #1497 is still open and failing. If someone who is more fluent in C than me could take a look, it would help us move closer to a fix.

**Comment by pitops** [comment]
@coaxial did you check latest comment in the PR ?

**Comment by coaxial** [comment]
> @coaxial did you check latest comment in the PR ?

Yes, didn't have the time to do it. Feel free to lend a hand.

**Comment by pitops** [comment]
> > @coaxial did you check latest comment in the PR ?
> 
> @pitops Yes, didn't have the time to do it. Feel free to lend a hand.

Unfortunately i have zero C experience

**Comment by coaxial** [comment]
I have rebased #1497 to `master` but I'm unable to do any further work on it. If anyone (or, who knows, Prusa themselves??) wants to carry on this work, knock yourself out!

**Comment by Prusa-Support** [comment]
This is a quite hectic time for our developer team.

You may have noticed a rather large log for the recent [pre-release](https://github.com/prusa3d/Prusa-Firmware-Buddy/releases/tag/v4.4.0-beta2
).
Probably you have also heard about other "big news" of announced products in the last year - which take up a little share of our developers' attention too.

Our developers will stick to a very tight and intense development plan for some time.

This doesn't mean your pull request will be forgotten. Your work is most valuable and truly appreciated, thanks.
Our devs will find time to dedicate to yours and the other pull requests as soon as possible. I hope you can understand.

Michele Moramarco
Prusa Research

**Comment by hphp67** [comment]
+1

**Comment by lukkigi** [comment]
+1 interested in this as well

**Comment by Prusa-Support** [comment]
No, I'm sorry, no news in this regard.
The news is that the ESP Wi-Fi module & Prusa Link implementation was successful and taken very positively by the customers.
Other than that, there is much more work going on under the hood - stay tuned.

The developers should find time to review this among a few more popular requests and I'm sorry I don't have an estimated time.

Michele Moramarco
Prusa Research

**Comment by arekm** [comment]
About M73 reporting status. MK3* firmware supports that fine for years. MK4 firmware doesn't support this at all. So now this should be considered a functionality regression between MK3 and MK4 in i3 line.

Fortunately M73 support is already in Marlin2 (2.1.2+) and Firmware-Buddy is based on Marlin2. Unfortunately based on older version but backporting that code looks to be easy enough. It's guarded by M73_REPORT, SET_PROGRESS_PERCENT, SET_REMAINING_TIME and SET_INTERACTION_TIME ifdefs in the Marlin2 repository code.

Backporting that would get Firmware-Buddy into MK3 firmware feature parity regarding M73 g-code.

Relevant code is mostly:
https://github.com/MarlinFirmware/Marlin/blob/0c2ed8688dd387b56b318f6207ee1cc70811b5a5/Marlin/src/gcode/lcd/M73.cpp#L58-L91

plus ui.get/set_XX() functions plus a bit more if M73 would have to affect also what's displayed on LCD.

**Comment by LordRaptor** [comment]
Please implement this, a lot of us are using Octoprint to control our printers and the standard set of G-Codes should be supported by all printers.

**Comment by hstag** [comment]
Gentlemen, this situation is unacceptable. Despite purchasing the Prusa Mini printer three years ago, the feature in question has not yet been implemented. Furthermore, it is disappointing to note that even the latest model of the printer, the MK4, lacks this feature, which represents a downgrade from the previous MK3s model.

**Comment by Prusa-Support** [comment]
Thanks for your solicitations, our developers and the whole company have great consideration for the community.
Our developers should stick to tight development plans at the moment but they will gladly look into this and other requests in the future. Until then, because the firmware is open source, I'm positive that the community's contribution can make a difference. 🙂  

For the records, the cameras are supported via Prusa Link (MK3 + Raspberry) or Prusa Connect.
Unfortunately, we still have a lack of documentation, but this was announced in the [release logs](https://github.com/prusa3d/Prusa-Link/releases), and here is a [list of supported cameras](https://github.com/prusa3d/Prusa-Link/blob/511dba9fd64a68f2362fe5355a67a024a7caede5/prusa/link/cameras/picamera_driver.py#L49). Alternatively, a "smart" camera device (smartphone, tablet, laptop...) can be connected via Prusa Connect very easily (QR code scan).

Michele Moramarco
Prusa Research

**Comment by LordRaptor** [comment]
I completely agree with Bpendragon. Having the Mk4 regress in capabilities vs the Mk3 seems a very weird choice. And it's not like you guys don't have a firmware that does all of this, just allocate a developer that worked on the Mk3 firmware to port these capabilities to the Mk4, or review the pull requests from the community that integrate this functionality and merge them. You keep claiming that your developers have to stick to tight development plans, but this is an issue that's been open for over 3 years now, development plans don't go that far into the future.

**Comment by fergymcferguson** [comment]
+1 Bump

**Comment by TravisWilder** [comment]
+1

**Comment by oxivanisher** [comment]
Please work on this issue. Buying the MK4 feels like a big downgrade process and ease of use wise which is really sad!

**Comment by Impatience71** [comment]
+1

I would also upvote this feature !

**Comment by rainer1146** [comment]
+1

**Comment by Soulofdestiny** [comment]
Yes please!

**Comment by Jackjan4** [comment]
+1 Bump.

**Comment by getgray** [comment]
Octoprint full compatibility is a need not a want.  Please implement ASAP

**Comment by ScottWell1** [comment]
+1

**Comment by Jinra** [comment]
Having a 3+ year old issue for this feature doesn't bode well for the MK4. +1 anyway

**Comment by TheNore** [comment]
Would it be possible to Please consider this a higher priority? As more mk4 units are being shipped more octoprint users are becoming frustrated and loosing functionality they have had in the MK3 printers for years!

**Comment by ghoeffner** [comment]
Wait until you find out that [the filament runout sensor doesn't work with octoprint anymore](https://github.com/prusa3d/Prusa-Firmware-Buddy/issues/3111)...

**Comment by fergymcferguson** [comment]
Nor can you see any of the files on the USB drive...

**Comment by beelsebob** [comment]
+1

**Comment by Ikario** [comment]
This is a must. I get that Prusa is interested in people using their Prusaconnect/Prusalink option but it is nowhere close to have the same features as Octoprint.

**Comment by jvasileff** [comment]
Pause print (`M601` added by PrusaSlicer) also doesn't work. I just built my MK4 yesterday, and I'm pretty disappointed that OctoPrint isn't really usable with it.

**Comment by yahbluez** [comment]
i second that.

**Comment by tideline3d** [comment]
24x MK3 farm here looking to expand soon. MK4 will have to be off the list for refusal to support 3rd party integrators with basic telemetry.  My entire production process is built around OctoPrint. 

I realize Prusa and Octoprint aren't on the best of terms but consider your customers rather than FOSS ecosystem politics here.

**Comment by FliesWithWind** [comment]
+1

**Comment by rnbarrett** [comment]
Count me in the upvote support!

**Comment by osbock** [comment]
Upvote for me too! I just ordered the Mk4 upgrade, and had I known that Octoprint didn't work as well as it does on the Mk3, I would not have purchased.  I want to support you, please support us!

**Comment by satchm0h** [comment]
+1

Please consider prioritizing support for the same interfaces/protocols/APIs the Mk2*->Mk3* supported. Your community have come to rely on many FOSS solutions that build upon these!

**Comment by SalmonSays** [comment]
+1 for interest, this feels like a massive oversight in MK4 release.

**Comment by shalmi** [comment]
@Prusa-Support PLEASE FIX THIS. It may not be important to you, but it IS important to your customers!

**Comment by Monkopedia** [comment]
@Prusa-Support +1 to everything here. I back ordered an mk4 a few days ago, but now must consider cancelling it given the lack of support here.

**Comment by kukulin** [comment]
+1

**Comment by goyetus** [comment]
Not a single answer from @michalxfanta  since 2020 .......   
This is not the way.......

**Comment by hstag** [comment]
Prusa support responded.
https://github.com/prusa3d/Prusa-Firmware-Buddy/issues/189#issuecomment-1585703918

**Comment by getgray** [comment]
I just cancelled my orders.  I'm seeing early adopters with issues, and Octoprint compatibility is a high priority for me.  I don't want to spend a grand to get upgrades that come with downgrades.

**Comment by jim-roberts** [comment]
+1 

 PrusaConnect/Link is not a suitable replacement.

**Comment by Monkopedia** [comment]
I just rebased @coaxial 's PR #1497 and have it building, however I don't have a suitable printer to try it on. So if anyone here is brave and wants to test #3147, please do. Otherwise, I guess I won't cancel my order, and hope we can manage to fix this ourselves.

**Comment by jvasileff** [comment]
@Monkopedia I cloned your repo, but then discovered there are no instructions for flashing custom firmware to the MK4. I asked Prusa support via chat if it is possible, and the response was "I don't think is possible at this point".

Also note that for the MINI, the instructions say:

> To flash your own (or a community-made) unsigned firmware, you’ll have to break the appendix on the logic board (called the Buddy board) and move the jumper in a correct position. Breaking the appendix on the Buddy board is irreversible and voids the warranty of your printer’s electronics.

So having a community fork won't help many people. We'll really need Prusa to help with this.

The earlier comment from Prusa Support isn't encouraging:

> Our developers should stick to tight development plans at the moment but they will gladly look into this and other requests in the future. Until then, because the firmware is open source, I'm positive that the community's contribution can make a difference.

Ironically, this regression with the MINI and MK4 actually cuts off a massive "community contribution" in the form of OctoPrint, which adds a ton of value to Prusa printers. I hope I'm wrong, but so far it looks like they aren't really interested in community contributions in this area, or they'd spend a little effort helping to fix the firmware for the huge gain of OctoPrint support.

**Comment by Ikario** [comment]
People expect this machines to improve upon their predecessors and have the same features + improvements.  If Octoprint compatibility was deprecated as a feature it should be clarified on their shop, but then again, the website advertises input shaping as a feature, so I'm not holding my breath on them either being up front about intentional lack of Octoprint support to push Prusaconnect/Prusalink or bringing that feature back any time soon.

**Comment by coaxial** [comment]
We've been making this point in this issue but also #283 (as the firmware doesn't support the "pause print" command either), or #1542 where the filament runout sensor still doesn't work with Octoprint. 

It's been over 3 years now with no technical help or input from anyone at Prusa (other than the support comments saying that basically it's an out-of-scope, nice-to-have feature)... 

Hopefully the fact that the MK4 has these issues too will bring more visibility to this issue as the Mini clearly isn't a priority at Prusa. 

This is very frustrating and disappointing overall, I will very seriously consider alternatives to Prusa when buying my next printer as a result.

**Comment by marek26340** [comment]
+1

**Comment by kaydeeH** [comment]
Not that Prusa seems interested in dealing with this, but I'll throw my own +1 into this. When my MK4 arrived and I found out about the almost complete lack of support for Octoprint on the new printers, I was flabbergasted. HUGE functionality to cut. Please add this functionality asap.

**Comment by alexhocks** [comment]
+1 from me as well. This feature is essential for anyone doing a bit more than the occasional hobby printing.

**Comment by fergymcferguson** [comment]
Bump. 

> Thanks for your solicitations, our developers and the whole company have great consideration for the community. Our developers should stick to tight development plans at the moment but they will gladly look into this and other requests in the future. Until then, because the firmware is open source, I'm positive that the community's contribution can make a difference. 🙂
> 
> For the records, the cameras are supported via Prusa Link (MK3 + Raspberry) or Prusa Connect. Unfortunately, we still have a lack of documentation, but this was announced in the [release logs](https://github.com/prusa3d/Prusa-Link/releases), and here is a [list of supported cameras](https://github.com/prusa3d/Prusa-Link/blob/511dba9fd64a68f2362fe5355a67a024a7caede5/prusa/link/cameras/picamera_driver.py#L49). Alternatively, a "smart" camera device (smartphone, tablet, laptop...) can be connected via Prusa Connect very easily (QR code scan).
> 
> Michele Moramarco Prusa Research

The developer of Octoprint has commented on the Prusa forums: “ No, "they" do not. Let me tell you about the reality of developing OctoPrint.

There's at this point what could be considered hundreds if not even thousands of firmware variants out there. Vendors randomly break things left and right. But do they tell me or anyone else? Of course not! We usually discover that something is broken because some affected users show up in our forums, Discord or bug tracker and tell us that something is not working like it should. If we are lucky, they do so in a nice tone and collaborate with us to figure out WHAT exactly is broken and if there's an easy way to work around it on OctoPrint's side (preferably through a plugin because every single workaround that makes its way into core now has to be maintained there for all eternity). Sadly all to often however, we rather see entitled passive aggressive stuff like also visible here, where people treat us like we owe them our time and resources to make THEIR specific workflow work for them, for free.

It is not my job to keep OctoPrint compatible to Prusa Research's Marlin derivatives just as it is not my job to keep OctoPrint compatible to any other printer vendor's firmware derivatives. We have an established communication protocol for the serial interface between hosts and printers. THAT is what I have to be compatible to, everything else I do my best to find the greatest common divisor for, so that implementing one feature for one derivative doesn't break stuff for everyone else. And if one vendor decides to go out of their way to do things differently, well, that is their decision. Protocol extensions are of course always an option, but that's something that needs to be documented and also needs to be discussed so it doesn't break stuff for everyone else.

So, to sum things up here - no, as always there has not been any kind of communication here (just as there wasn't any for other developments in the past that then cost me an arm and leg in user support overhead). That is not my decision, that is Prusa Research's decision. My job is to keep OctoPrint working and keep improving it for hundreds of thousands of users out there, and only a fraction of those have a MK4 or another Prusa printer to begin with. Things work flawlessly for people who do not have a MK4, so whatever the issue is here, it very much sounds like it's a firmware issue specific to the MK4 firmware flavour. As always, I'm happy to assist the firmware devs in figuring out what's wrong if I get tagged in a public ticket somewhere (THAT is something I consider part of my job, within sensible limits), and if something more in depth is needed I'm also open to contracting work upon request, but that's all I can offer given my responsibilities towards ALL of OctoPrint's user base here.”

It sounds to me like the community is trying to contribute but have been unable to make any progress because of Prusa’s developers. C’mon guys - 2 years of this is outright ignoring what your users are asking for. Either flip the switch to enable the feature which is already available on MK3’s or let us flash our own firmware without voiding our warranty.

**Comment by ghoeffner** [comment]
^ Here is the original post: https://forum.prusa3d.com/forum/postid/662700/

**Comment by dMAC95** [comment]
I build printrun based farm software for a company that runs 100+ MK3s. We had a sizeable MK4 order placed and also received an early pre-built MK4 to run preliminary testing on. When we found out the MK4 didn't support host communications like the MINIs and MK3s do and also came to realize prusa wasn't prioritizing the development of it we canceled our order. I tried building custom firmware, but that's out of my league so now we have a spare MK4 collecting dust with a cut appendix.

**Comment by Monkopedia** [comment]
@dMAC95 Any chance you're looking to offload that mk4 to someone who is willing to do custom firmware development?

**Comment by dMAC95** [comment]
> @dMAC95 Any chance you're looking to offload that mk4 to someone who is willing to do custom firmware development?

I'm not sure how to DM on github, but we may be interested. If you can build firmware to support octoprint I'll be sure that our business interests don't get in the way of it being released publicly.

**Comment by dMAC95** [comment]
> 

Also I'd be interested in learning the basics of building that kind of firmware. In time I'd love to jump in on your Buddy pull request. I have a web application background, but I can figure things out with the right resources.

**Comment by fergymcferguson** [comment]
This is the third oldest and most commented open feature request in this repo. 

If you’re stalling to drive people to PrusaConnect/PrusaLink, I think the only thing you’re gong to accomplish is to drive people to other brands. How unfortunate when the community at large is already saying Prusa is falling behind the curve of innovation and this makes the newest flagship printer a downgrade for many.  

PrusaLink is a joke in its current state and PrusaConnect is not much better. If you want people to use Prusa’s solution for remote/farm management, it needs to be way better. Prusa is clearly limited in what they can do in terms of development since you’re sticking to your “tight schedule” and in the mean time users have to suffer. Octoprint works great because any dev can make plugins to add functionality instead of waiting for an official release from Prusa to add even the most basic features like a legitimate camera stream which vanilla Octoprint does. 

I wish some of the prominent social media content creators would pick up this thread and start shining some light on this issue and maybe then it will get some traction.

**Comment by jvasileff** [comment]
> so now we have a spare MK4 collecting dust with a cut appendix.

@dMAC95 do you have any more info about the appendix? I see where it is on the board, but is there also a jumper that needs to be set? I only found official docs for the mini.

**Comment by dMAC95** [comment]
> > so now we have a spare MK4 collecting dust with a cut appendix.
> 
> @dMAC95 do you have any more info about the appendix? I see where it is on the board, but is there also a jumper that needs to be set? I only found official docs for the mini.

I only had to cut the bridge to get it working. I also came across the same resources as you describing moving a board component to a different connection as with the MINI, but the MK4 board is similar to the MK3. I remember when I opened the MK4 the appendix bridge was within the two slot holes near to the top left of the board while it was still upright within the housing. I didn't need to remove the board to cut the appendix.

**Comment by Ikario** [comment]
> This is the third oldest and most commented open feature request in this repo.
> 
> If you’re stalling to drive people to PrusaConnect/PrusaLink, I think the only thing you’re gong to accomplish is to drive people to other brands. How unfortunate when the community at large is already saying Prusa is falling behind the curve of innovation and this makes the newest flagship printer a downgrade for many.
> 
> PrusaLink is a joke in its current state and PrusaConnect is not much better (and I’m really trying to give it a fair shot here). If you want people to use Prusa’s solution for remote/farm management, it needs to be way better. Prusa is clearly limited in what they can do in terms of development since you’re sticking to your “tight schedule” and in the mean time users have to suffer. Octoprint works great because any dev can make plugins to add functionality instead of waiting for an official release from Prusa to add even the most basic features like a legitimate camera stream which vanilla Octoprint does.
> 
> I wish some of the prominent social media content creators would pick up this thread and start shining some light on this issue and maybe then it will get some traction.

I could not agree more.  What makes it more infuriating is that this is not just some random requested new feature, this was something that was actually removed.  I can tolerate (though I don't agree and I think is very shady) advertising input shaping and pressure advance when none of those features are actually working as they should (an alpha release is not a stable release and crash detection does not work with input shaping at all, another feature removed if you want to get the capabilities you paid for), but actively removing features to push for a propietary system that is extremely slow (115200bps max), basic and feature incomplete is, to say it lightly, not in the spirit of openness and to be real, very anti-competitive. I read this a few months ago on reddit and had not realized how right this person was up until now:
"Remember a while ago when Josef Prusa sent out a request for people to vote on the things they wanted to see most on the MK4? Well, he took the top 10 of those and added them to the MK4 sales brochure just before he released it. He hasn't added them to the MK4 yet, though."
The strategy seems to be getting input shaping on the mk4 firmware ready first, then focusing on Prusalink/Prusaconnect to get feature parity with Octoprint so that people will drop the subject but everyone that follows prusa development knows this is in no way shape or form realistic because they don't usually do well with schedules and there are features on the mini, around 3 years ago, that were promised and they never got around to it.  This is a really big issue that needs to be addressed as it severely handicaps the MK4 that is already arguably an inferior product compared to the competition at this same price point.

**Comment by jvasileff** [comment]
@dMAC95 @Monkopedia I removed the appendix and attempted to install custom firmware. But it failed to boot. I'll share exactly what I tried in #3147

@Prusa-Support In your last message, you punted development on this issue to the community. Can you at least provide firmware build & installation instructions for the MK4?

**Comment by mikolaszuza** [comment]
I am not a dev, so take my info with a grain of salt. I asked about this in our dev channel and they said fixing stuff such as the filament runout sensor or progress on the LCD screen when running octo is definitely on the todo list. Unfortunately I can't give a timeline right now.

I wouldn't even say it's about giving priority to PrusaConnect over this, that's an entirely different team working on that. Of course, the fw and PrusaConnect team do have to interact. But on the fw side of things, I think the prio right now is still improving the network stack, which even octo users should end up benefiting from, even if just partially.

**Comment by Ikario** [comment]
> 

Mikolas,
I get that this is not your area, so first of all I'd like to thank you for showing up here and giving some info about what's going on behind the scenes. It's the second response we've seen in 3+ years from Prusa and the first one was... not so great, let's just put it that way.
But I do kinda have to push back on that a bit, because this has been a thing for 3+ years at this point, so they might be improving the network stack right now and I get that, it's great! but this must have been pushed down the prio a TON if it hasn't seen any kind of progress whatsoever in 3 years.  Again, I'd like to stress that this is REMOVED functionality, if right now you buy the latest Prusa printer, you stand to lose features that the previous version already had. I would believe restoring said features so that basic functionality is there should have some kind of priority.  Again, I know this is not your thing (and I really appreciate you coming forward) but hey, it's still is a thing.

**Comment by mikolaszuza** [comment]
Totally get the frustration with issues open this long. I disagree that this is removed functionality, we didn't remove this from the firmware. We didn't somehow port all our 8-bit features to 32-bit platform. We had to write them all again for the entire new OS architecture. But I can understand how it may seem that way looking at the MK3 vs MK4 when it comes to octo support.

**Comment by Ikario** [comment]
Yeah, I feel like that's arguing semantics, you did not delete the code that made everything work, but by not porting the part that made octoprint work, as you said, the end result is pretty much exactly the same.  An MK3 has something that was considered a given and is now NOT on the MK4, something that anyone upgrading would not expect to lose.  And as for speculating about being done on purpose to push PrusaLink/Connect. No, I don't actually believe it was done in bad faith but then again, every response we've had so far before you from Prusa was in fact dismissive of the issue AND pushed PrusaLink/Connect as a replacement.

**Comment by jvasileff** [comment]
I don't believe the firmware was totally rewritten from scratch. From what I can tell, it would be more accurate to say that new or reworked parts of the firmware broke some existing features and code.

But regardless, rather than splitting hairs about the software implementation, from a user's perspective, the feature has been removed when going from MK3 to MK4. Consumers do not review GitHub repositories before making purchasing decisions.

I was quite surprised to learn that USB printing isn't supported on the MK4 when I tried to use it after finishing building my kit. It seemed so obvious to me that Prusa would maintain compatibility with open solutions rather than push towards proprietary lock-in, and this assumption informed my purchasing decision when comparing against Prusa's new competition.

**Comment by dMAC95** [comment]
I'm sure that porting firmware from 8-bit to 32-bit is not a straightforward process. I believe people are frustrated at the lack of communication and the release of the MK4 without the expected features. It would have been better for Prusa to release the MK4 after ironing out the firmware instead of releasing it half baked. I want to ensure that we don't take our frustration out on @mikolaszuza for sharing these details and hope to keep communication open.

**Comment by ddanier** [comment]
@mikolaszuza First of all thanks for your reply, very much appreciated!

I just want to add one aspect to this removed or not removed feature aspect. From my perspective it is true the MK4 did not really remove a MK3 feature here - at the end of the day the firmware is completely new. Still I understand how people can feel like this, as it’s just absolutely unexpected behavior.

But then the MK4 firmware is based on what you have done for the Mini, they both use the firmware buddy software. If I compare the OctoPrint/USB printing on those two the Mini has at least a printing screen where you can change things like some print settings. Also this mode switch in the firmware is responsible for the filament sensor to trigger, which also is currently missing when printing using OctoPrint on MK4 (meaning prints fail miserably when the firemen runs out + you have to go through some extended cleaning process as the filament is then stuck after the extruder). The Mini print screen is far from perfect and certainly not what people are used from the MK3, but still the Mini has some support for this.

So at some point on moving the firmware from the Mini to the MK4 this functionality seems to have gone away. Looking at it from this perspective it clearly is a feature that has been removed (or just not yet ported to the new screen dimensions?).

For me getting back what has been working on the Mini would solve about 80% of this issue. Also there must be some code already doing this on the Mini…I would love to see this being made compatible with the MK4 as a first step.

**Comment by coaxial** [comment]
@ddanier Mini support with Octo has also been very underwhelming ever since the Mini was released. The filament sensor sporadically works, pause, play a sound aren't supported over serial (they're just not implemented, which is curious because Marlin supports these codes), we've had very underwhelming official answers for the last 3+ years on these issues.  Same thing for showing progress messages on screen...

I'm glad MK4 users have joined the party as we might finally be getting some attention from Prusa now, but as a Mini owner I feel pretty neglected, dismissed, and unheard for the last few years. These issues have been there from day one, this bait and switch is a running joke. 

Prusa: this is not acceptable. We paid full price for the printers, we are owed fully functional firmware (including a full implementation when using serial/Octoprint) and you've had over 3 years to look into this.

**Comment by marek26340** [comment]
I have to agree with coaxial there.

Mini's support for Octoprint has been atleast working fine for me just printing. I don't have a filament sensor on mine, so can't judge on anything there.
I'd like to ask - what do you really need the beep function in gcode for? And when was the last time you seriously needed that?
And print progress? I take a quick look either right into the Octoprint GUI or at the printer and take my guess. It's finished when it's finished, 3D printers usually take their sweet time and there's not much for you to do to hurry it up (likely not without hurting quality). And you should let the bed cool down before taking the finished print off the bed anyway. I'm not trying to argue or dispute anything here, please take these words with a grain of salt.

> Impatience can cause wise people to do foolish things. 
\- Janette Oke

Anyway. Ever since firmware 4.4 for the Mini, the connection with Octoprint is extremely buggy and pretty much unusable. MBL sends incomplete data to Octoprint, preventing the Bed Visualizer plugin from working correctly (I use it to make sure my Mini is properly aligned/square).
And I also had a weird problem after upgrading. Looks like the accelerations or something is broken in the firmware because any motion of the axis now sounds like the printer is about to die. Sounds like sticking bearings, loud snaps and stuff like that. Print quality also went down. Not too much, but noticeable for me.
Downgrading back down to 4.3.4 fixed everything. (#3042, #3036 and more)

**Comment by goyetus** [comment]
Thanks a lot @mikolaszuza!  for trying to get this problems to Joseph  and Dev Team.   

I am very disappointed in the way @josefprusa has answered. 1 second later he changed to "it looks like some people are interested in TPU".

The problem with octorprint is NOT SMALL. A lot of people depend on octoprint to use their machines.

We have already heard several times that there are other priorities ahead, like making the NETWORK go faster (since 2 years you have it pending on the MINI which happens to use the same WIFI module as the MK4 and which CASUALLY has a speed of 128.000 baud....... i.e. you are passing the WIFI network through the COM. Until you stop doing that and write directly to the SD card / pendrive, you will continue with the same 10-20 kbps !!!!!

PLEASE, FIX OCTOPRINT. Filament Sensor is really  important.  
M600  and another marlin gcodes must be a priority  (at least in my opinion)

**Comment by coaxial** [comment]
> I am very disappointed in the way @josefprusa has answered. 1 second later he changed to "it looks like some people are interested in TPU".

Either no one sees the problem, or they're not interested. 

Either way, it doesn't look good for those here who want Octoprint feature parity.

**Comment by Ikario** [comment]
> Thanks a lot @mikolaszuza!  for trying to get this problems to Joseph  and Dev Team.   
> 
> I am very disappointed in the way @josefprusa has answered. 1 second later he changed to "it looks like some people are interested in TPU".
> 
> The problem with octorprint is NOT SMALL. A lot of people depend on octoprint to use their machines.
> 
> We have already heard several times that there are other priorities ahead, like making the NETWORK go faster (since 2 years you have it pending on the MINI which happens to use the same WIFI module as the MK4 and which CASUALLY has a speed of 128.000 baud....... i.e. you are passing the WIFI network through the COM. Until you stop doing that and write directly to the SD card / pendrive, you will continue with the same 10-20 kbps !!!!!
> 
> PLEASE, FIX OCTOPRINT. Filament Sensor is really important. M600 and another marlin gcodes must be a priority (at least in my opinion)

Yeah, I would agree that in this case no mention might have been even better than what we got, it looks like it's so far down the to do list that maybe we'll get there a few months after the MK4S comes out. Maybe.

**Comment by arekm** [comment]
> I'm sure that porting firmware from 8-bit to 32-bit is not a straightforward process.

Note that Prusa didn't port its firmware from 8-bit to 32-bit. Prusa 32-bit firmware was done in this way: take Marlin 2 firmware - marlinfw.org - (ancient version as of today), disable almost all features from it and only leave what Prusa thinks is okish and then add own Prusa stuff to that.

So for example current version of Marlin 2 has all these features that people report as missing (by missing I mean functionality that was in 8bit fw but is not available in 32bit fw). But Prusa-Firmware-Buddy doesn't update from Marlin 2. Prusa Firmware Buddy is just a permanent fork done at few years ago.  Since that it is going with own path.

Current list of missing functionality (that I know) is listed below. My knowledge state is about 4.6 firmware versions but I guess nothing new happened in these areas. All these things work fine in MK3S+ (8bit firmware) and are not working in new 32bit firmware (Mini, MK4, XL).

- no information is printed on LCD for prints from octoprint
- M600
- no more support for timestamps and long filenames when listing medium (USB) - https://github.com/prusa3d/Prusa-Firmware-Buddy/issues/3054
- M23 doesn't work (start print from usb medium) - https://github.com/prusa3d/Prusa-Firmware-Buddy/issues/2387 
- M73 doesn't report back to host (to octoprint) remaining print time, time to color change. Breaks various octoprint plugins.

**Comment by goyetus** [comment]
Font forget :
- M412  .   Filament Runout.     Mk4 is not sending any code when it runs out filament.    Very dangerous this one,  and the main reason of my worry.

**Comment by LordRaptor** [comment]
Wait the MK4 does not support M600, even when printing from USB?

Now I am seriously thinking about sending back my unopened Mk4 kit. Who knows how many month it will take for them to implement this. It really feels like we were all sold an unfinished product. Prusa lost a lot of goodwill with this

**Comment by ddanier** [comment]
M600 works if you send it via USB. After having used my nearly empty spools on my Mini exclusively lately cause of the missing filament runout handling I just did add a Bigtreetech Smart Filament Sensor to my setup 🙈. When OctoPrint sends the M600 it works like normal now.

If you send M600 manually it also just works.

Still this is something Prusa needs to fix asap. The situation is really bad, as this is a very basic printer requirement. Having to fix what Prusa „forget“ to implement is just not acceptable. 🧐

**Comment by LordRaptor** [comment]
That sounds like M600 works, so what about the command doesn't work?

**Comment by ddanier** [comment]
@mikolaszuza Any news on this? 😬

**Comment by goyetus** [comment]
@mikolaszuza 
Any update in basic commands like m600 ?   any ETA?

**Comment by arichens** [comment]
Leaving my +1 here. As a maker who has multiple brands of printers, I have no plans to ever adopt PrusaLink/PrusaConnect. Full Octoprint support with M412/M600, etc. is a deal breaker for myself and many others.

**Comment by goyetus** [comment]
I will sell my mk4 if is not fixed before 2024....  

I can not wait 2 years like the Mini  with no support.

**Comment by mtaumike** [comment]
> I will sell my mk4 if is not fixed before 2024....
> 
> I can not wait 2 years like the Mini with no support.

Drastic but I totally understand.

I have zero want or need of the Prusa(Link|Connect) ecosystem

**Comment by goyetus** [comment]
> > I will sell my mk4 if is not fixed before 2024....
> > I can not wait 2 years like the Mini with no support.
> 
> Drastic but I totally understand.
> 
> I have zero want or need of the Prusa(Link|Connect) ecosystem

^The problem is noy only prusaLink..... is the lack of all the octoprint plugins,    the slow wifi at 14 kb/s  like the Prusa Mini ,  the lack of "camera support"  (oh ,   you can use a mobile anyways...) ....   It´s like a bad joke.

**Comment by getgray** [comment]
After waiting months, just before it it might of shipped (it was very late already), I cancelled my MK4 order over this issue.  I'll consider placing another when they get Octoprint compatibility sorted.  But from the looks of it the MK4 may be outdated, superseded, or have a better alternative by then.  We'll see.

**Comment by mikolaszuza** [comment]
I just got back from a developer meeting about the firmware schedule.
The priority on this issue was bumped up in the previous meetings, everyone is well aware of how important this is for you all. It will likely be in the 5.1 firmware, ETA 2 months. What's almost finished is G-code streaming, binary G-code encoding, and arc-welding (these combined create massive file transfer speedups), those features will also be in this firmware release. 

Sharing things just heard in meetings is always scary because unexpected problems can come up and timelines can slip (I am inflating the deadline they told me to better manage everyone's expectations). But this is clearly an important task to communicate updates on, so here I am, please don't kill the messenger (me). 

If I hear any major updates on this (positive or negative), I'll try to update here again.

**Comment by oxivanisher** [comment]
> I just got back from a developer meeting about the firmware schedule. The priority on this issue was bumped up in the previous meetings, everyone is well aware of how important this is for you all. It will likely be in the 5.1 firmware, ETA 2 months. 

Thank you so much! Hearing that it is officially planned for a future release is great! <3

> If I hear any major updates on this (positive or negative), I'll try to update here again.

Communication is the key and as long as it look like it is worked on, I am waaaay happier!

THANKS!

**Comment by TravisWilder** [comment]
> Yes, though the different printers, MK4, XL, and MINI, might be updated at somewhat different times and with some differences due to hardware limitations.

any news on the 5.0 as well?

**Comment by goyetus** [comment]
thanks a lot @mikolaszuza !!!!

**Comment by ddanier** [comment]
Thank you very very much @mikolaszuza! Very nice to get some progress and updates on this.

I can only talk for myself, but I won't blame you if something happens and the timeline slips - as long as you are open about it and keep us in the loop. This is still way better than having no idea about what happens and ones option is only to hope for things to change. Thanks again for shedding some light this issue.

**Comment by tkircher** [comment]
> I just got back from a developer meeting about the firmware schedule. The priority on this issue was bumped up in the previous meetings, everyone is well aware of how important this is for you all. It will likely be in the 5.1 firmware, ETA 2 months. What's almost finished is G-code streaming, binary G-code encoding, and arc-welding (these combined create massive file transfer speedups), those features will also be in this firmware release.

 This is great news. I've been using arc welder with my MK2.5S for every print since I found out it existed. Significant improvement in print quality in addition to smaller gcode, and arcs would be a good special case for input shaping on the MK4.

 A little offtopic but I'm planning to use the Pi 4's OTG function to emulate a USB storage device, to try to circumvent some of the issues with the MK4 relating to octoprint. It can't solve the filament runout reporting issue but it can make files available to both the printer and octoprint at the same time.

**Comment by jim-roberts** [comment]
> > I just got back from a developer meeting about the firmware schedule. The priority on this issue was bumped up in the previous meetings, everyone is well aware of how important this is for you all. It will likely be in the 5.1 firmware, ETA 2 months. What's almost finished is G-code streaming, binary G-code encoding, and arc-welding (these combined create massive file transfer speedups), those features will also be in this firmware release.
> 
> 
> 
>  This is great news. I've been using arc welder with my MK2.5S for every print since I found out it existed. Significant improvement in print quality in addition to smaller gcode, and arcs would be a good special case for input shaping on the MK4.
> 
> 
> 
>  A little offtopic but I'm planning to use the Pi 4's OTG function to emulate a USB storage device, to try to circumvent some of the issues with the MK4 relating to octoprint. It can't solve the filament runout reporting issue but it can make files available to both the printer and octoprint at the same time.

FYI you don't need Octoprint to use arc welder. You can add arc welder as part of your g code creation workflow by including an invocation in Prusaslicer to the local instance of the arc welder library. https://github.com/FormerLurker/ArcWelderLib

**Comment by tkircher** [comment]
> FYI you don't need Octoprint to use arc welder. You can add arc welder as part of your g code creation workflow by including an invocation in Prusaslicer to the local instance of the arc welder library. https://github.com/FormerLurker/ArcWelderLib

I assumed when they said they'd be supporting it in firmware, and that it would reduce upload time, it meant they didn't have arcs implemented in the gcode interpreter. If you send it to the printer and it does the conversion locally, it wouldn't affect the upload time.

**Comment by mtaumike** [comment]
> I assumed when they said they'd be supporting it in firmware, and that it would reduce upload time, it meant they didn't have arcs implemented in the gcode interpreter. If you send it to the printer and it does the conversion locally, it wouldn't affect the upload time.

Well, if you do use ArcWelder from within PS it should reduce the file size of the resultant gcode which should shorten the time to upload.

The printer is not going to convert into G2/G3 arcs for ya

**Comment by Ikario** [comment]
> > Well, if you do use ArcWelder from within PS it should reduce the file size of the resultant gcode which should shorten the time to upload.
> > The printer is not going to convert into G2/G3 arcs for ya
> 
> I think you misunderstood. It was stated that "G-code streaming, binary G-code encoding, and arc-welding" would be supported in the later _firmware_. Meaning that it isn't supported now. If that's the case, then the MK4 may not support arcs at all, meaning using arcwelder in prusaslicer and uploading that gcode would just result in a confused printer.

This is 100% wrong.  The MK4 firmware supports G2/G3 commands. I have used ArcWelder and I'm printing gcode with G2/G3 commands while typing this message.  Also, PrusaSlicer (I believe from 2.60 onwards) can visualize gcode with G2/G3 correctly.  Also, before people ask, this works with both regular and IS profiles.

**Comment by tkircher** [comment]
> This is 100% wrong. The MK4 firmware supports G2/G3 commands. I have used ArcWelder and I'm printing gcode with G2/G3 commands while typing this message. Also, PrusaSlicer (I believe from 2.60 onwards) can visualize gcode with G2/G3 correctly. Also, before people ask, this works with both regular and IS profiles.

Thanks for clearing that up! In that case I'm not sure what supporting arc-welder in firmware would mean, other than actually processing the gcode and adding arcs.

**Comment by mikolaszuza** [comment]
"Octoprint is not working the way we want on the MINI / MK4 / XL (e.g. no progress on the LCD screen, not being able to swap filament...). The problem is on our side (not in Octoprint). Our firmware team is working on resolving these issues and improving Octoprint integration.

The fix will be part of the upcoming 5.1 firmware. Estimated time to release is 6 weeks."
https://x.com/Prusa3D/status/1705151177329959185?s=20

**Comment by bdynamic** [comment]
I really look forward for the improvement. The Display of by good old MK2.5 had more useful information displayed than the shiny MK4 ;-) Otherwise: Great device - I love it!

**Comment by foorschtbar** [comment]
> I just got back from a developer meeting about the firmware schedule. The priority on this issue was bumped up in the previous meetings, everyone is well aware of how important this is for you all. It will likely be in the 5.1 firmware, ETA 2 months. What's almost finished is G-code streaming, binary G-code encoding, and arc-welding (these combined create massive file transfer speedups), those features will also be in this firmware release.
> 
> Sharing things just heard in meetings is always scary because unexpected problems can come up and timelines can slip (I am inflating the deadline they told me to better manage everyone's expectations). But this is clearly an important task to communicate updates on, so here I am, please don't kill the messenger (me).
> 
> If I hear any major updates on this (positive or negative), I'll try to update here again.

[FW 5.1](https://github.com/prusa3d/Prusa-Firmware-Buddy/releases/tag/v5.1.0-alpha2) alpha is out, but "Due to the major changes to the code, the Octoprint is currently not working, however, we are already working on the fix."

Will it be fixed in the final 5.1 firmware?

**Comment by mikolaszuza** [comment]
Update from FW 5.1 Beta https://github.com/prusa3d/Prusa-Firmware-Buddy/releases/tag/v5.1.0-beta
> Alpha support of Octoprint
> 
> This release includes improvements for users utilizing OctoPrint for remote printing. It's important to note that this implementation is still a work in progress, some features will be added (fixed) in future releases.
> 
> Current OctoPrint limitations
> 
>     Pause/Resume commands from OctoPrint are not functional (Pause/Resume the print using the screen for now).
>     The Stop print command from OctoPrint is not operational (Stop the print using the screen for now).
>     Multitool XL - If the tool crashes, the print cannot be recovered.

**Comment by Bpendragon** [comment]
> Update from FW 5.1 Beta https://github.com/prusa3d/Prusa-Firmware-Buddy/releases/tag/v5.1.0-beta
> 
> > Alpha support of Octoprint
> > This release includes improvements for users utilizing OctoPrint for remote printing. It's important to note that this implementation is still a work in progress, some features will be added (fixed) in future releases.
> > Current OctoPrint limitations
> > ```
> > Pause/Resume commands from OctoPrint are not functional (Pause/Resume the print using the screen for now).
> > The Stop print command from OctoPrint is not operational (Stop the print using the screen for now).
> > Multitool XL - If the tool crashes, the print cannot be recovered.
> > ```

I haven't had time to test it, but how does Octoprint handle binary GCode?

I know it can handle arc moves fine, but the new binary format is a huge question mark (I have no issue with just reverting to normal gcode for the time being though)

**Comment by dirkvranckaert** [comment]
> ###
> **[mikolaszuza](/mikolaszuza) ** commented [Nov 20, 2023](#issuecomment-1818998327)

My first print on this alpha started just now. Nice to see the OctoPrint logo on the MK4's display :-)
However this first print is a _safe_ one that should finish without issues. After that I have print waiting that needs filament change at about 75% because the roll will run out, hopefully this time the MK4 will properly pause and let me change the filament!

![IMG_7333](https://github.com/prusa3d/Prusa-Firmware-Buddy/assets/4660794/99f27c38-5ea6-4d93-9de1-12b36efb83f9)

**Comment by tkircher** [comment]
The octoprint logo is a nice touch ^_^

**Comment by hstag** [comment]
I do not think the logo is so nice if there is no progress shown. This logo was implemented from the beginning on the mini, and until now there was no will implement the progress bar. This was no effort to implement it on mk4. On mk3 we had progress displayed.

**Comment by dirkvranckaert** [comment]
Not having the progress bothers me less... Would be nice indeed if that gets added, but personally don't need it since the print farm is managed entirely through OctoPrint / OctoFarm
@mikolaszuza Is there any view atm on which features will be _ported_ or which not such as the progress @hstag is talking about?

**Comment by tkircher** [comment]
> I do not think the logo is so nice if there is no progress shown. This logo was implemented from the beginning on the mini, and until now there was no will implement the progress bar. This was no effort to implement it on mk4. On mk3 we had progress displayed.

It actually never occurred to me that you'd want a progress bar on the display. All I really want is Octoprint to be fully integrated so I can print using it. For a file that takes several minutes to upload using the builtin ESP wifi module, or ethernet, it is effectively instantaneous with a RPi. The Prusa web interface just barely works for printing, while Octoprint has dozens of plugins I like to use and supports a webcam, all with gcode synchronization and UI feedback. I'd give up the display entirely if I could have that working.

**Comment by ScottWell1** [comment]
The "progress" display with Octoprint is a part of standard, basic Marlin M117 support.  In my opinion, there is simply NO EXCUSE for Prusa's Buddy firmware omitting M117 support.  Whether YOU need it or not, it is part of baseline features that should be included in any Marlin branch.

**Comment by coaxial** [comment]
> All I really want is Octoprint to be fully integrated so I can print using it.

This is moving the goal posts. Octoprint support is table stakes when it comes to 3D printers nowadays, much more so when it's a major, premium brand's flagship model. It should have been there from day one instead of the half broken Prusa connect thing to nobody ever asked for. 

> M117

Same for this code, it comes with Marlin and is a basic, standard feature. How is it that 4+ years after the Mini (aka the "print farm works horse") has been released and months after the MK4 is out we *still* don't have support for such a basic feature and have to beg/lobby Prusa to do anything about it in threads like this? 

Had MK4 users not felt the pain of missing features in the firmware, I'm not sure anything would have moved for existing Mini owners. We've been asking for these features for years and barely ever got any attention.

**Comment by hstag** [comment]
I FULLY agree with @coaxial. This should be done **before** releasing a mk4 and mini too. I sold my mk3's and bought mk4 because I thought the new version will be superior and will be an upgrade. I made a mistake.

And as an owner of mini I am really disappointed. 

Begging and waiting years to implement things that was promised in the beginning is really not a respectable way of treating the customers.

**Comment by ddanier** [comment]
Just a note: You can view the progress if you go to "Tune" -> "Message History" (last element in menu list). This shows a log of messages which should include progress and current layer.

I know this is not optimal and I myself wish the integration would be better. However you can look at the progress this way, which is still a good thing. ;-)

**Comment by coaxial** [comment]
I'm not sure if this has been fixed, but using this menu to watch progress made the filament runout sensor not trip anymore when running out of filament.

**Comment by dirkvranckaert** [comment]
> Just a note: You can view the progress if you go to "Tune" -> "Message History" (last element in menu list). This shows a log of messages which should include progress and current layer.
> 
> I know this is not optimal and I myself wish the integration would be better. However you can look at the progress this way, which is still a good thing. ;-)

Indeed, saw that menu too but some basic info should simply be available on screen immediately...

**Comment by ddanier** [comment]
One nice thing I just noticed (and tested on MK4): The "Tune" menu is now much more in sync with what is possible during USB printing. You for example can use "Cancel object" while printing using OctoPrint. Also the LED status bar matches what is happening + this also shows up in ProseConnect. Pretty neat.

Now I'm waiting for some print job details (progress, current/total layer, ETA)....as this is what is issue is originally all about.

**Comment by Prusa-Support** [comment]
For not natively integrated features - or not yet implemented - please consider opening new issues.
I recommend opening issues with a clear title and description for each single request.
This allows for better tracking of specific requests and bug reports in general.

Michele Moramarco
Prusa Research

**Comment by coaxial** [comment]
Done: https://github.com/prusa3d/Prusa-Firmware-Buddy/issues/3653

**Comment by hstag** [comment]
It's great that after 4 years you opened a new issue with the same description instead of implementing it. 

4 years.

**Comment by timur-tabi** [comment]
See also See also #3846.

**Comment by Silvenga** [comment]
Bump.

**Comment by nealzip** [comment]
I'd like to upvote this.

**Comment by jerekarsten** [comment]
I guess it's easier to attach a 16X2 lcd display to my raspberry to monitor printing progress like it's 2018, as this apparently is impossible to do with the firmware. They put a man on the moon you know?

**Comment by timur-tabi** [comment]
Let's be perfectly clear, this is not a hard problem.  Prusa could do it easily.  Heck, they could have an intern do it.  They've had years to fix this problem.  No, the reason this problem exists is because _Prusa does not care_.  It's just not important enough to them.

**Comment by tkircher** [comment]
The MK3 had hardware support for a Pi Zero running Octoprint and it looked like Prusa was committed to freedom. At some point since then, they must have had some very bad advice from a consultant, and decided to close the platform and commit to a proprietary lock-in like Bambu Labs has. There was obviously huge backlash and I think they did realize they made a mistake but it may take a very long time for them to fully support an open platform.

**Comment by Prusa-Support** [comment]
> Prusa could do it easily.

At this point, probably literally anyone can do it, in a similar fashion they do for Octoprint.

> The MK3 had hardware support for a Pi Zero running Octoprint and it looked like Prusa was committed to freedom

We may have a different idea of "freedom" and I hope freedom is granted on both ends or maybe I'm not entirely sure what you mean.
Yes, we tried to support Octoprint compatibility more actively from our side in the past and we know it takes special maintenance and attention but this is not all! Certain features that are almost entirely based on Octorpint are not directly compatible unless the printer's firmware is substantially modified and thoroughly tested.
Other than "freedom" I would like to talk about Open Source which seems to be the only bridge between our company and Octoprint. Prusa is a huge supporter of Open Source, in fact, everyone can contribute to the project and feel _free_ to customize the firmware to their liking, for themselves and/or for the community.

We will continue monitoring and making considerations.
Thanks for your trust in the Open Source and for your understanding.

Michele Moramarco
Prusa Research

**Comment by timur-tabi** [comment]
> we tried to support Octoprint compatibility more actively from our side in the past and we know it takes special maintenance and attention but this is not all! 

We're not asking for Octoprint support, we're asking you to support the serial connect that is part of your hardware.  The giant Octopus that you display on the screen, instead of useful data, is completely unnecessary and no one asked for that.  You invented that all on your own.  What we want is for you to restore the same functionality that you used to deliver in your previous product.  That's it.

**Comment by Ikario** [comment]
This whole post says literally nothing when it's all about standard basic
functionality that was removed and replaced with a silly picture. I don't
know what this rant is about other than saying "we couldn't care less".
Disappointing and a shame. At least it's signed so we know who we can't
rely on to significantly improve user experience on prusa printers, thank
you for that.

**Comment by tkircher** [comment]
> > The MK3 had hardware support for a Pi Zero running Octoprint and it looked like Prusa was committed to freedom
> 
> We may have a different idea of "freedom" and I hope freedom is granted on both ends or maybe I'm not entirely sure what you mean.

 It means you commit to supporting open source tools, not your own inferior proprietary solution (Prusa Connect) which takes away freedom. And I'm not sure what you mean by both ways. Prusa is a corporation, not a person, and should have no freedom. Its only purpose is to do what customers want.

**Comment by ScottWell1** [comment]
> Blah blah blah...
> **_We will continue monitoring and making considerations._**  
> Blah blah blah...

This entire post is one of the most insulting missives I have ever seen any company employee make towards their customers. If anyone spoke to customers of my company this way, they would be looking for a new job.

And calling it "Octoprint support" is either deliberately disingenuous or shows a huge amount of ignorance.  What is being asked for are standard two-way communication support for Serial Print control -- commands like M117, which have been COMMONPLACE in virtually every printer manufactured in the last decade -- **including Prusa's own MK3 line.**  Those are not "unique" to Octoprint -- rather, they are needed for any serial print controller to function well with Prusa's products.

Stop gaslighting and telling us to "go support ourselves."  Prusa printers come at a premium price, and all we want is for it to live up to "standard" support -- so please quit making excuses and support your customers with the features they need.

**Comment by kbranch** [comment]
The fact that the Prusa "solution" (PrusaLink) is so, so much worse than Octoprint makes this much more frustrating. I don't care about Octoprint on my Klipper machine because Klipper is powerful and open (both in terms of source and communication protocol). Prusa has had 5 years to either add support for a couple simple gcodes that the MK3 supported or develop an alternative that's any good.

Prusa is struggling hard right now to keep up with its competitors in terms of hardware and value. Support for the community and easy interoperability are two of the few strong points they can still claim. Both the tone and content of this response are disappointing.

**Comment by lucasoskorep** [comment]
> Certain features that are almost entirely based on Octorpint are not directly compatible unless the printer's firmware is substantially modified and thoroughly tested. 

Standards exist for a reason.  As has been stated - nobody is asking you to support Octoprint specific features or rewrite your entire firmware.  Its one thing to have an "open source" product with all contributions funneled through a for profit company, and another entirely to actually embrace open source and support open standards.  

The ask on this thread is official support for 2 small open standards in M73 (print progress) and M117 (text updates).

**Comment by Sainx47** [comment]
Any update on supporting M73 and M117 ?

**Comment by coaxial** [comment]
Celebrating a fifth year of waiting for this feature soon 🎉

**Comment by jdfalk** [comment]
The only way they are going to get it is if enough people vote with their wallets and make this painful for them. In my off time I work with several school districts and I've already stopped recommending Prusa printers for this exact reason. I won't even consider telling anyone else to get one because of the insulting behavior from their staff, and complete disregard for standards. 

I do wonder if we are confusing them by referencing octoprint all the time instead of just saying we want these gcodes over serial for any piece of software to communicate with. They seem to be hung up on the specifically supporting octoprint part, where we want this because it's part of the standard that they purposely stripped out, or; I'd hate to be as condescending and insulting as prusa support was, but maybe it's just too confusing and we should use smaller words for these middle management types. How about this: "me user, you have dev, me need gcode, you add gcode. Ok?"

This will probably get removed but someone had to say it.

**Comment by nstivachtis** [comment]
I just upgraded from a mk3 to an XL and I'm pretty annoyed by some of the functionality loss when using octoprint and this is one of them.  I wouldn't mind if prusaLink was a useful product.  And I have no love for cloud based...  Gien the length of time this request has been open for it's clear that team Prusa just doesn't want to support M73 and M117 anymore and would rather have you locked into their cloud...  I wonder if they're somehow monetizing useahe data

**Comment by DarkPhyber-hg** [comment]
I just upgraded to an XL from a mk3 which worked great with octoprint.  Prusalink is a joke for those of us not wanting our printer on the cloud.  Even if Prusalink got to be better, it would still need to support something similar to self-hosted Obico for me to consider using it.

If Prusa wanted to enable M73 and M117, I'm sure they easily could have in the 5 years this issue has been open.  As a consumer nothing is more frustrating than to see an issue open this long, with this many people commenting and voting on it, and for the company to just not do anything about it.  Especially when it is functionality they removed from a previous release.  Other enhancements that may require bigger changes I would understand, but 2 g-code commands seems so trivial.  

In my opinion Prusa doesn't want to do this because they want us to use their cloud product, I'm sure they mine a lot of valuable data from Prusa Connect. 

Very disappointed in Prusa, and regretting my decision to get my XL.  You'd think Prusa would do more for the community given they are having such a hard time keeping up with the likes of the Bambu x1, which based on early reviews still seems to be a better printer and better value than the Core One.  Not to mention the other value based core xy printers that are just hitting the market.  Once another tool changer printer comes out, I'm really not sure why anyone would go with a Prusa product anymore because this thread just makes me feel like Prusa just doesn't care about the community.

**Comment by getgray** [comment]
These and some related Octoprint issues, I’m fuzzy in this one. My MK3 doesn’t show much on the printer screen for an Octoprint initiated print. Is that all these 2 codes do?  Show information on the printers screen?  As long as I can see what’s going on in Octoprint I can live with no screen. OR if this issue prevents the printer from working with Octoprint at all, then I’m out.

**Comment by Jinra** [comment]
@getgray IIRC when I had my MK3 with octoprint, it would show the information it normally does as if it was printing from the printer directly such as time remaining/time remaining until filament change, and other info. With MK4 and octoprint, it shows an octoprint icon with the option to pause or cancel

**Comment by williamoverstreet** [comment]
+1

**Comment by jneilliii** [comment]
I think that the most important part to reiterate here is that this feature request is not an OctoPrint specific support request. It's a gcode support request for **any** host connected software sending commands over serial. What I'm confused about is the reprap wiki lists M73 as supported for Buddy.

https://reprap.org/wiki/G-code#M73:_Set.2FGet_build_percentage

![Image](https://github.com/user-attachments/assets/b6484041-895f-41a4-bce4-56b81247bd1d)

Same for M117

https://reprap.org/wiki/G-code#M117:_Display_Message

![Image](https://github.com/user-attachments/assets/db18a080-50d3-4143-a455-597a4fcaa8a7)

Has something changed and this issue is still just hanging around or does the feature only work for files printing from SD/USB/PrusaConnect?

I agree with the sentiment that others here have posted, it's strange for the most reacted and largest number of replies for a feature request to not have any movement in 5+ years.

**Comment by ruedli** [comment]
I hope it is still considered. "Just" displaying an Octoprint logo and not the information that is contained in M117 messages is a waste of space in the user-interface.

**Comment by CZDanol** [comment]
Yes, we haven't thrown the idea out of the window. But sorry, octoprint/serial support is not very high in the priority list.

**Comment by foorschtbar** [comment]
> Yes, we haven't thrown the idea out of the window. But sorry, octoprint/serial support is not very high in the priority list.

Then I would think about it again. There are people who buy Prusa printers because, unlike Bambulab, you can print here without the cloud and don't lose any convenience with OctoPrint. That's why I don't understand why Prusa try to replicate everything with PrusaPrint, PrusaLink and EasyPrint. There is a hard and loyal core of people who buy Prusa because of OctoPrint support and we are getting more and more annoyed.

**Comment by tkircher** [comment]
The 'cloud' just means no privacy, no control, and general inconvenience. So, Prusa Connect never should have existed. And PrusaLink is pointless because OctoPrint already supports standard printers. For some reason Prusa removed the already existing display support to try to force people to use Prusa Connect when nobody ever wanted that, and they must have known that intentionally disabling those commands would make OctoPrint harder to use in particular.

**Comment by redPanther** [comment]
just a note from a p mini user ... for various reasons, I had to print via prusa connect/link and not octo as usual. I must say, yeah it looks nicer, but it lacks display of current layer number. Just the percentage and time when it's ready is not so usefull for me.

I use it for triggering filament change during print of a layer (do it mostly when I bored of mod the gcode) to have e.g. not the whole layer in the new color.

So I'm happy when I can switch back to my ugly but informative octo print screen. I wish prusa could activate the message window when serial printing. This is definitely not a big deal! @CZDanol would be nice if the devs can auto switch to messages when serial printing and ditch the nice octo picture.
I guess lot's of users complaining didn't know that the can view the messages from octoprint on the display and that octo can show usefull stuff there (e.g. print time, current/total layer, % of the print)


@ruedli 
|    I hope it is still considered. "Just" displaying an Octoprint logo and not the information that is contained in M117 messages is a waste of space in the user-interface.

during print: press "tune" and then messages. then you see all your messages comming from octoprint.

**Comment by ruedli** [comment]
This is good to know, it shows at the same time that te information is
readily available, so not showing it and choose for a the smiling octopus
is indeed a waste of UI space.

I wonder if it can be done as a customization, there is no pull request
with implementation yet?

Ruud Rademaker
prefered mail: ***@***.***


Op wo 25 jun 2025 om 12:42 schreef redPanther ***@***.***>

**Comment by redPanther** [comment]
@ruedli BTW I mentioned this in this thread some time ago https://github.com/prusa3d/Prusa-Firmware-Buddy/issues/189#issuecomment-873104604 
I also added ETA via DisplayLayerProgress plugin (not shown on this pic0)

and to put more salt in the soup, there is a pr that perhaps needs some work, but implements important stuff for octoprint support ( https://github.com/prusa3d/Prusa-Firmware-Buddy/pull/3976 ) so developers don't have to start from scratch

**Comment by Jinra** [comment]
I have my Core One built with Octoprint and was very confused why every print suddenly seemed frozen with the firmware responding `Recv: echo:busy: processing` in Octoprint's terminal. This was not conveyed in any way in either octoprint or the screen, so I had no way to know what was wrong. It wasn't until I uploaded to prusa link (which takes way too long to do) that I discovered that the printer was doing an "Absorbing Heat" step for 7 minutes prior to printing.

It'd be nice to have _any_ kind of telemetry on the screen so people don't spend needless time diagnosing things.

**Comment by reboot81** [comment]
I'd also appreciate if this got some attention.

**Comment by Zwirbel1** [comment]
I am also waiting for an update on this.

**Comment by Dawied** [comment]
> Yes, we haven't thrown the idea out of the window. But sorry, octoprint/serial support is not very high in the priority list.

I prefer serial printing above all other options (cloud, wifi). Please let us know if it could be bumped up on the priority list (progress info and thumbnail).

**Comment by bkerler** [comment]
A lot of things have been changed in the past firmware versions regarding the display and gui, so this request isn't forgotten. I will have a closer look at the implementation and I'll see if I can do anything to make this happen. Sorry for the long wait, we will keep you guys updated !

**Comment by jkotowicz123** [comment]
Please do - even M117 fixes in alpha 6.4.0 still dont apply to octoprint - M117 message is still not displayed, nor is job progress or file name. Basic information like this is really needed, as it was on MK3/S for years. 

Although if there was a way to give some access to the display via serial, it would be really handy. For example we would love to be able to dynamically set image in place of octoprint logo. We could display thumbnails from octoprint or icons/ warnings for farm operators about nozzle size or hardware malfunctions. Now we need to use stickers, because AFAIK, there is zero access to that display. 

And PrusaConnect is still way,way underpowered for us. I am afraid even with this upcoming "ERP like" version that Joseph teased at Fab25 it still might be too "close-source" for our needs, so we need decent octoprint support.

**Comment by bkerler** [comment]
M117 is a bit difficult to implement. It does require a change of the existing fsm handler as this current isn't able to forward char arrays but only values, at least this was the case I had a look at it previously.

**Comment by azizcontractor** [comment]
+1 on this. Any support would be appreciated. I’m in a similar boat where I can’t see the absorbing heat whenever I use octoprint

**Comment by msandy** [comment]
> there is a pr that perhaps needs some work, but implements important stuff for octoprint support ( https://github.com/prusa3d/Prusa-Firmware-Buddy/pull/3976 ) so developers don't have to start from scratch

@redPanther That's a fairly massive PR. Are you able to highlight the specific changes related to handling M73 and M117?

**Comment by foorschtbar** [comment]
> highlight the specific change

@msandy https://github.com/prusa3d/Prusa-Firmware-Buddy/pull/5071/changes

**Comment by ttabi** [comment]
A pull request was submitted four months ago.  The Prusa developers should review it.

**Comment by jneilliii** [comment]
Literally the oldest open issue with the most comments with a pending Pull Request. 

<img width="854" height="263" alt="Image" src="https://github.com/user-attachments/assets/3a53a391-b4b5-4a75-bacb-de936360b6fc" />

**Comment by tkircher** [comment]
> Literally the oldest open issue with the most comments with a pending Pull Request.

Yeah, but Prusa doesn't want to admit that nobody wants to use their 'ecosystem'. :D

**Comment by ruedli** [comment]
What makes it especially awkward is the many times Prusa promised to look into it. Seems a waste of time to never finish it.
